### PR TITLE
Fix definite assignment analysis issue caused by `x !== null` checks

### LINKED ISSUE #49386: Definite assignment checking confused by `x !== null` check
### 🕗 Version & Regression Information

Nightly (4.8.0-dev.20220603)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220603#code/GYVwdgxgLglg9mABMOcA8AVAfACgB4BciGiAPomCADZUCUiA3gFCKuJUCmUiAnogLyI8AbhZsYwRDj4BCfoMo16zNqt7DWAek2IAogCd9cfTLGsAvk0tA)

### 💻 Code

```ts
function foo<T>(x: T | null) {
    let y = x;
    if (y !== null) {
        y;  // Error!
    }
}
```

### 🙁 Actual behavior

Error

### 🙂 Expected behavior

No error

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #49386.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
Issue was originally discovered in #49343.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -18164,6 +18164,10 @@ namespace ts {
             return false;
         }
 
+        function containsUndefinedType(type: Type) {
+            return !!((type.flags & TypeFlags.Union ? (type as UnionType).types[0] : type).flags & TypeFlags.Undefined);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:18168]: We have to be careful with this - nothing with typeflags other than `Unknown` and `Any` can now precede the group of `Undefined`-ish types starting with `undefinedType` until `missingType`.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test this
@typescript-bot perf test faster

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts:18168]: Agreed, but we already do this in a few places in the interest of performance.

**Comment by ahejlsberg** [comment]
@DanielRosenwasser Tests all look clean and performance is unaffected. I think this one is good to go!

**Pull request merged**